Add temporary directory support to _WKWebExtensionControllerConfiguration.

xeenon opened the pull request:
#### f1350da694fea03048cedf20314aaff4bc9b28d9
<pre>
Add temporary directory support to _WKWebExtensionControllerConfiguration.
<a href="https://webkit.org/b/266041">https://webkit.org/b/266041</a>
<a href="https://rdar.apple.com/problem/119346018">rdar://problem/119346018</a>

Reviewed by Brian Weinstein.

Adds a way to run extensions with temporary storage for testing persistence features.
Also removes the caching in createStorageDirectoryPath() which would cache the path
with the identifier or default path and never return a different value.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
(+[_WKWebExtensionControllerConfiguration configurationWithIdentifier:]):
(+[_WKWebExtensionControllerConfiguration _temporaryConfiguration]):
(-[_WKWebExtensionControllerConfiguration encodeWithCoder:]):
(-[_WKWebExtensionControllerConfiguration initWithCoder:]):
(-[_WKWebExtensionControllerConfiguration debugDescription]):
(-[_WKWebExtensionControllerConfiguration _isTemporary]):
(-[_WKWebExtensionControllerConfiguration _storageDirectoryPath]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfigurationPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
(WebKit::WebExtensionControllerConfiguration::createStorageDirectoryPath): Fixed incorrect caching.
(WebKit::WebExtensionControllerConfiguration::createTemporaryStorageDirectoryPath): Added.
(WebKit::WebExtensionControllerConfiguration::copy const):
(WebKit::WebExtensionControllerConfiguration::declarativeNetRequestStoreDirectory):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.cpp:
(WebKit::WebExtensionControllerConfiguration::WebExtensionControllerConfiguration):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
(WebKit::WebExtensionControllerConfiguration::createTemporary): Added.
(WebKit::WebExtensionControllerConfiguration::storageIsTemporary const): Added.
(WebKit::WebExtensionControllerConfiguration::storageDirectory const): Added.
(WebKit::WebExtensionControllerConfiguration::setStorageDirectory): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionControllerConfiguration.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager initForExtension:]):
(-[TestWebExtensionManager initForExtension:extensionControllerConfiguration:]):
(TestWebKitAPI::Util::loadAndRunExtension):

Canonical link: <a href="https://commits.webkit.org/271741@main">https://commits.webkit.org/271741@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84b3bdcdc38c52c63d636ba3ca3cf21df87bb00c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29368 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8035 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30697 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31933 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26663 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29968 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5330 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26660 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29641 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6789 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25122 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5745 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5947 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26178 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33274 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26900 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26611 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32100 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5846 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4031 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29883 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7549 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6371 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3796 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6370 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
```diff
@@ -149,6 +175,18 @@ - (void)setWebViewConfiguration:(WKWebViewConfiguration *)configuration
+
+- (NSString *)_storageDirectoryPath
+{
+    if (auto& directory = _webExtensionControllerConfiguration->storageDirectory(); !directory.isEmpty())
```
didn't know you could do this all in one line! very cool

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
```diff
@@ -96,7 +93,7 @@
     }
 
-    m_declarativeNetRequestStoreDirectory = FileSystem::pathByAppendingComponent(m_storageDirectory, "DeclarativeNetRequest"_s);
+    m_declarativeNetRequestStoreDirectory = FileSystem::pathByAppendingComponent(storageDirectory(), "DeclarativeNetRequest"_s);
```
any particular reason to use `storageDirectory()` over `m_storageDirectory` since the method doesn't make any additional changes?

xeenon replied:
It is usually best to use getters, incase there is extra logic in them. It is inline, so it is just the same in this case. But it is more future proof.

xeenon committed:
Add temporary directory support to _WKWebExtensionControllerConfiguration.
https://webkit.org/b/266041
rdar://problem/119346018

Reviewed by Brian Weinstein.

Adds a way to run extensions with temporary storage for testing persistence features.
Also removes the caching in createStorageDirectoryPath() which would cache the path
with the identifier or default path and never return a different value.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
(+[_WKWebExtensionControllerConfiguration configurationWithIdentifier:]):
(+[_WKWebExtensionControllerConfiguration _temporaryConfiguration]):
(-[_WKWebExtensionControllerConfiguration encodeWithCoder:]):
(-[_WKWebExtensionControllerConfiguration initWithCoder:]):
(-[_WKWebExtensionControllerConfiguration debugDescription]):
(-[_WKWebExtensionControllerConfiguration _isTemporary]):
(-[_WKWebExtensionControllerConfiguration _storageDirectoryPath]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfigurationPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
(WebKit::WebExtensionControllerConfiguration::createStorageDirectoryPath): Fixed incorrect caching.
(WebKit::WebExtensionControllerConfiguration::createTemporaryStorageDirectoryPath): Added.
(WebKit::WebExtensionControllerConfiguration::copy const):
(WebKit::WebExtensionControllerConfiguration::declarativeNetRequestStoreDirectory):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.cpp:
(WebKit::WebExtensionControllerConfiguration::WebExtensionControllerConfiguration):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
(WebKit::WebExtensionControllerConfiguration::createTemporary): Added.
(WebKit::WebExtensionControllerConfiguration::storageIsTemporary const): Added.
(WebKit::WebExtensionControllerConfiguration::storageDirectory const): Added.
(WebKit::WebExtensionControllerConfiguration::setStorageDirectory): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionControllerConfiguration.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager initForExtension:]):
(-[TestWebExtensionManager initForExtension:extensionControllerConfiguration:]):
(TestWebKitAPI::Util::loadAndRunExtension):

Canonical link: https://commits.webkit.org/271741@main
